### PR TITLE
feat: Add JSON serialization capabilities

### PR DESCRIPTION
**Pull request by FrogTheFrog**
## Description
Added JSON serialization capabilities for custom types using the NLOHMAN JSON library.
The implementation details (library) are not exposed by design! As a side effect, people will have to
declare which types they want to use.

The main purpose of JSON is for logging mostly and in one case it will be used for persistence (future PR).

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

File: cmake/Boost_DD.cmake [hunk 1]
```diff
@@ -2,10 +2,9 @@
 # Loads the boost library giving the priority to the system package first, with a fallback
 # to the submodule.
 #
-include_guard(GLOBAL)
-
-find_package(Boost 1.85)
+find_package(Boost 1.85 QUIET)
```
Comment (ReenigneArcher) [inline cmake/Boost_DD.cmake:5]: Why quiet?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -46,6 +46,12 @@ endfunction()
 add_subdirectory(fixtures)
 add_subdirectory(unit)
 
+#
+# Additional external libraries
+#
+include(Boost_DD)
+include(Json_DD)
+
```
Comment (ReenigneArcher) [inline tests/CMakeLists.txt:54]: Doesn't this automatically pick up the parent includes?

File: cmake/Boost_DD.cmake (hunk 1 above)
Comment (FrogTheFrog) [reply cmake/Boost_DD.cmake:5]: If the package is not found it prints messages whenever the file is included. These are annoying, especially when there no version of the said package in the system as the warnings are quite long.

I thought quiet option is acceptable in this case as we are printing our custom message once anyway...

File: tests/CMakeLists.txt (hunk 2 above)
Comment (FrogTheFrog) [reply tests/CMakeLists.txt:54]: find_package does not add packages globally I think, while other methods do.

There's a GLOBAL flag that's available in 3.24 that I would like to use though and get rid of this in the future...

File: cmake/Boost_DD.cmake (hunk 1 above)
Comment (ReenigneArcher) [reply cmake/Boost_DD.cmake:5]: Sounds fine to me, thanks!

File: tests/CMakeLists.txt (hunk 2 above)
Comment (ReenigneArcher) [reply tests/CMakeLists.txt:54]: In Sunshine, I just bumped one of our docker images from 3.18 to a the latest cmake, so it shouldn't be too hard to do that if there's a few others lagging.

Comment (FrogTheFrog) [reply tests/CMakeLists.txt:54]: Does it mean I should bump CMake to 3.24 now and remove these?

Comment (ReenigneArcher) [reply tests/CMakeLists.txt:54]: You could if you'd like to

**Pull request merged**